File: Week 01/problems/arrays/reverse_array_groups.cc
#include <bits/stdc++.h>
using namespace std;

class Solution{
	public:
	    void reverseInGroups(vector<long long>& arr, int n, int k){
	        if (k == 1) return;
	        if (k >= n){
	        	reverse(arr.begin(), arr.end());
	        }
	        else {
	        	for(int i = 0; i < n; i += k){
	            int start = i;
	            int end = min(i+k-1, n-1);
	            
		            while (start < end){
		                swap(arr[start], arr[end]);
		                start++;
		                end--;
		            }
	        	}
	        }
	    }
};

int main() {
    int t; 
    cin >> t; 
    while(t--){ 
        int n;
        cin >> n; 
        vector<long long> arr; 
        int k;
        cin >> k; 

        for(long long i = 0; i<n; i++)
        {
            long long x;
            cin >> x; 
            arr.push_back(x); 
        }
        Solution ob;
        ob.reverseInGroups(arr, n, k);
        
        for(long long i = 0; i<n; i++){
            cout << arr[i] << " "; 
        }
        cout << endl;
    }
}

File: Week 01/problems/arrays/reverse_array_groups.py
# https://practice.geeksforgeeks.org/problems/reverse-array-in-groups0255/1/
# https://www.geeksforgeeks.org/reverse-an-array-in-groups-of-given-size-2/
# https://www.geeksforgeeks.org/reverse-an-array-in-groups-of-given-size/

class Solution:
	# time is O(n) | space is O(1)
	def reverse_in_groups(self, arr, N, K):
		if K == 1: 
			return

		if (K >= N):
			arr.reverse()
		else:
			for i in range(0, len(arr), K):
				start = i
				# to handle case when k is not multiple of n 
				end = min(i+K-1, N-1) # since k is const

				while (start < end):
					arr[start], arr[end] = arr[end], arr[start]
					start += 1
					end -= 1

import math
def main():
    T=int(input())
    while(T>0):
        nk=[int(x) for x in input().strip().split()]
        N=nk[0]
        K=nk[1]
        arr=[int(x) for x in input().strip().split()]
        
        ob = Solution()
        ob.reverse_in_groups(arr,N,K)
        for i in arr:
            print(i,end=" ")
        print()
        T-=1

if __name__=="__main__":
    main()
